testing/src/test/java/io/stargate/it/driver/CqlSessionExtension.java

```diff
@@ -0,0 +1,405 @@
+ *
+ * <p>A {@code USE ks} statement is automatically issued, so that the session is already "connected"
+ * to the keyspace when your test code starts executing. In general you don't need to prefix table
+ * names in your queries, and tests are often oblivious to the actual keyspace name. However it can
+ * be injected if you need it (see the "Parameter injection" section below).
+ *
+ * <h3>Customization</h3>
+ *
+ * Certain aspects of the session creation can be controlled with the {@link CqlSessionSpec}
+ * annotation. See the javadocs of that type for more details.
+ *
+ * <h3>Parameter injection</h3>
+ *
+ * The extension will detect and inject the following parameters on JUnit test methods
+ * (constructors, BeforeXxx or test methods, etc):
+ *
+ * <ul>
+ *   <li>any parameter of type {@link CqlSession}: injected with the session that was created by the
+ *       extension.
+ *   <li>any parameter of type {@link CqlSessionBuilder}: injected with a new session builder
+ *       pre-configured to connect to the Stargate cluster. This allows you to create new sessions
+ *       with a different configuration in your test methods. Note that such sessions are not
+ *       managed by the extension: <b>you must close them explicitly once you are done</b>, to avoid
+ *       resource leaks. A try-with-resources block is generally the best way to do that. Also, you
+ *       might want to invoke {@link #awaitAllNodes(CqlSession, StargateEnvironmentInfo)} if there
+ *       are multiple Stargate nodes.
+ *   <li>a parameter of type {@link CqlIdentifier} <b>AND annotated with {@link TestKeyspace}</b>:
+ *       injected with the identifier of the keyspace that was created by the extension. This is
+ *       useful if you need to look up schema metadata, use the keyspace from another session or
+ *       API, etc.
+ * </ul>
+ */
+public class CqlSessionExtension
+    implements BeforeAllCallback,
+        BeforeEachCallback,
+        AfterEachCallback,
+        AfterAllCallback,
+        ParameterResolver {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CqlSessionExtension.class);
+  private static final int KEYSPACE_NAME_MAX_LENGTH = 48;
+
+  // Tracks whether the extension was configured per-class or per-method.
+  private volatile Lifecycle sessionLifecycle;
+
+  private volatile StargateEnvironmentInfo stargate;
+  private volatile CqlSessionSpec cqlSessionSpec;
+  private volatile CqlIdentifier keyspaceId;
+  private volatile CqlSession session;
+
+  @Override
+  public void beforeAll(ExtensionContext context) {
+    LOG.debug("Starting per-class execution for {}", context.getElement());
+    // beforeAll and afterAll are never called if we're per-method, that's how we figure out the
+    // lifecycle
+    sessionLifecycle = Lifecycle.PER_CLASS;
+    createSession(context);
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) {
+    if (sessionLifecycle == null) {
```

Comment:
If the test method has `@ExtendWith(CqlSessionExtension.class)` then we'll get a new instance of this class for each test method, AFAIK, so checking instance variables may not be reliable... WDYT?

Reply:
It may be preferable to store session data in the `ExtrensionContext.Store`... :thinking: